server/workflow/workflow_server.go

```diff
@@ -20,14 +20,25 @@ import (
 	"github.com/argoproj/argo/workflow/validate"
 )
 
+// RunningMode is the way that workflowServer is used, GRPC server side or kube client.
+type RunningMode string
+
+// Running modes
+const (
+	KubeClient RunningMode = "KubeClient"
```

Comment:
how can we get this to work with both clients?

Reply:
`argo-server-client` does grpc call to argo server, it doesn't run this code on the client side. 

So there are only 2 places running this, `argo-kube-client` and `GRPC server`.

Reply:
ok, so if KubeClient needs to run server code, we should refactor that code out

Reply:
Then what is the reason for change https://github.com/argoproj/argo/pull/2290? It was this change to remove all the classic client side service code and reuse the server code. @alexec

Reply:
yes - as most of the logics are in the server - but I don't understand why the server needs to know if it is in server more or kube client - it is not IOC

Reply:
See my comments at `withInstanceID()`.